File: app/src/main/java/com/gytxtx/openjbd/protocol/JbdDeviceInfo.java
package com.gytxtx.openjbd.protocol;

public final class JbdDeviceInfo {
    public static final JbdDeviceInfo EMPTY = new JbdDeviceInfo("", "", "", "", "", "", 0f, 0f, 0f);

    public final String serialNumber;
    public final String barcode;
    public final String batteryModel;
    public final String manufacturer;
    public final String bmsModel;
    public final String bmsAddress;
    public final float ratedChargeCurrentA;
    public final float ratedDischargeCurrentA;
    public final float ratedDischargePowerW;

    private JbdDeviceInfo(
            String serialNumber,
            String barcode,
            String batteryModel,
            String manufacturer,
            String bmsModel,
            String bmsAddress,
            float ratedChargeCurrentA,
            float ratedDischargeCurrentA,
            float ratedDischargePowerW
    ) {
        this.serialNumber = clean(serialNumber);
        this.barcode = clean(barcode);
        this.batteryModel = clean(batteryModel);
        this.manufacturer = clean(manufacturer);
        this.bmsModel = clean(bmsModel);
        this.bmsAddress = clean(bmsAddress);
        this.ratedChargeCurrentA = ratedChargeCurrentA;
        this.ratedDischargeCurrentA = ratedDischargeCurrentA;
        this.ratedDischargePowerW = ratedDischargePowerW;
    }

    public JbdDeviceInfo withSerialNumber(String value) {
        return new JbdDeviceInfo(value, barcode, batteryModel, manufacturer, bmsModel, bmsAddress, ratedChargeCurrentA, ratedDischargeCurrentA, ratedDischargePowerW);
    }

    public JbdDeviceInfo withBarcode(String value) {
        return new JbdDeviceInfo(serialNumber, value, batteryModel, manufacturer, bmsModel, bmsAddress, ratedChargeCurrentA, ratedDischargeCurrentA, ratedDischargePowerW);
    }

    public JbdDeviceInfo withBatteryModel(String value) {
        return new JbdDeviceInfo(serialNumber, barcode, value, manufacturer, bmsModel, bmsAddress, ratedChargeCurrentA, ratedDischargeCurrentA, ratedDischargePowerW);
    }

    public JbdDeviceInfo withManufacturer(String value) {
        return new JbdDeviceInfo(serialNumber, barcode, batteryModel, value, bmsModel, bmsAddress, ratedChargeCurrentA, ratedDischargeCurrentA, ratedDischargePowerW);
    }

    public JbdDeviceInfo withBmsModel(String value) {
        return new JbdDeviceInfo(serialNumber, barcode, batteryModel, manufacturer, value, bmsAddress, ratedChargeCurrentA, ratedDischargeCurrentA, ratedDischargePowerW);
    }

    public JbdDeviceInfo withBmsAddress(String value) {
        return new JbdDeviceInfo(serialNumber, barcode, batteryModel, manufacturer, bmsModel, value, ratedChargeCurrentA, ratedDischargeCurrentA, ratedDischargePowerW);
    }

    public JbdDeviceInfo withRatings(float chargeCurrentA, float dischargeCurrentA, float dischargePowerW) {
        return new JbdDeviceInfo(serialNumber, barcode, batteryModel, manufacturer, bmsModel, bmsAddress, chargeCurrentA, dischargeCurrentA, dischargePowerW);
    }

    public boolean hasAnyField() {
        return serialNumber.length() > 0
                || barcode.length() > 0
                || batteryModel.length() > 0
                || manufacturer.length() > 0
                || bmsModel.length() > 0
                || bmsAddress.length() > 0
                || ratedChargeCurrentA > 0f
                || ratedDischargeCurrentA > 0f
                || ratedDischargePowerW > 0f;
    }

    private static String clean(String value) {
        if (value == null) {
            return "";
        }
        return value.replace('\u0000', ' ').trim();
    }
}

File: app/src/main/kotlin/com/gytxtx/openjbd/protocol/JbdDeviceInfo.kt
package com.gytxtx.openjbd.protocol

class JbdDeviceInfo(
    @JvmField val serialNumber: String,
    @JvmField val barcode: String,
    @JvmField val batteryModel: String,
    @JvmField val manufacturer: String,
    @JvmField val bmsModel: String,
    @JvmField val bmsAddress: String,
    @JvmField val ratedChargeCurrentA: Float,
    @JvmField val ratedDischargeCurrentA: Float,
    @JvmField val ratedDischargePowerW: Float
) {
    companion object {
        @JvmField
        val EMPTY: JbdDeviceInfo = JbdDeviceInfo("", "", "", "", "", "", 0f, 0f, 0f)
    }

    fun withSerialNumber(value: String) = copy(serialNumber = value)

    fun withBarcode(value: String) = copy(barcode = value)

    fun withBatteryModel(value: String) = copy(batteryModel = value)

    fun withManufacturer(value: String) = copy(manufacturer = value)

    fun withBmsModel(value: String) = copy(bmsModel = value)

    fun withBmsAddress(value: String) = copy(bmsAddress = value)

    fun withRatings(chargeCurrentA: Float, dischargeCurrentA: Float, dischargePowerW: Float) =
        copy(
            ratedChargeCurrentA = chargeCurrentA,
            ratedDischargeCurrentA = dischargeCurrentA,
            ratedDischargePowerW = dischargePowerW
        )

    fun copy(
        serialNumber: String = this.serialNumber,
        barcode: String = this.barcode,
        batteryModel: String = this.batteryModel,
        manufacturer: String = this.manufacturer,
        bmsModel: String = this.bmsModel,
        bmsAddress: String = this.bmsAddress,
        ratedChargeCurrentA: Float = this.ratedChargeCurrentA,
        ratedDischargeCurrentA: Float = this.ratedDischargeCurrentA,
        ratedDischargePowerW: Float = this.ratedDischargePowerW
    ) = JbdDeviceInfo(
        serialNumber, barcode, batteryModel, manufacturer, bmsModel, bmsAddress,
        ratedChargeCurrentA, ratedDischargeCurrentA, ratedDischargePowerW
    )

    fun hasAnyField(): Boolean =
        serialNumber.isNotEmpty()
                || barcode.isNotEmpty()
                || batteryModel.isNotEmpty()
                || manufacturer.isNotEmpty()
                || bmsModel.isNotEmpty()
                || bmsAddress.isNotEmpty()
                || ratedChargeCurrentA > 0f
                || ratedDischargeCurrentA > 0f
                || ratedDischargePowerW > 0f
}
